File: Groak_iOS/Groak/Groak/RestaurantObjects/Table.swift
//
//  Table.swift
//  Groak
//
//  Created by Vishisht Tiwari on 11/6/19.
//  Copyright © 2019 Groak. All rights reserved.
//
//  Table class used to refer to a table

import Firebase

// Different table status of a table
internal enum TableStatus: String {
    case available
    case seated
    case ordered
    case approved
    case served
    case payment
};

internal class Table {
    var name: String
    var reference: DocumentReference?
    var originalReference: DocumentReference?
    var orderReference: DocumentReference?
    var requestReference: DocumentReference?
    var restaurantReference: DocumentReference?
    var status: TableStatus
    var newRequest: Bool
    var serveTime: Timestamp
    
    init() {
        name = ""
        reference = nil
        orderReference = nil
        originalReference = nil
        requestReference = nil
        restaurantReference = nil
        status = TableStatus.available
        newRequest = false
        serveTime = Timestamp.init()
    }
    
    init(table: [String: Any]) {
        name = table["name"] as? String ?? ""
        reference = table["reference"] as? DocumentReference
        orderReference = table["orderReference"] as? DocumentReference
        originalReference = table["originalReference"] as? DocumentReference
        requestReference = table["requestReference"] as? DocumentReference
        restaurantReference = table["restaurantReference"] as? DocumentReference
        status = (table["status"] as? String).map { (TableStatus(rawValue: $0) ?? TableStatus.available) } ?? TableStatus.available
        newRequest = table["newRequest"] as? Bool ?? false
        serveTime = table["serveTime"] as? Timestamp ?? Timestamp.init()
    }
    
    func success() -> Bool {
        if (name.count == 0) {
            return false;
        } else if (reference == nil) {
            return false;
        } else if (originalReference == nil) {
            return false;
        } else if (orderReference == nil) {
            return false;
        } else if (requestReference == nil) {
            return false;
        } else if (restaurantReference == nil) {
            return false;
        }
        return true;
    }
    
    var description : String {
        var str = "Table Name: \(name)\n"
        str += "Table Reference: \(reference?.documentID ?? "Table reference not present")\n"
        str += "Table Original Reference: \(originalReference?.documentID ?? "Table  original reference not present")\n"
        str += "Order Reference: \(orderReference?.documentID ?? "Order reference not present")\n"
        str += "Requests Reference: \(requestReference?.documentID ?? "Requests reference not present")\n"
        str += "Restaurant Reference: \(restaurantReference?.documentID ?? "Restaurant reference not present")\n"
        str += "Table Status: \(status)"
        str += "New Request: \(newRequest)"
        str += "Serve Time: \(serveTime)"
        
        return str;
    }
}

File: Groak_android/app/src/main/java/com/groak/groak/restaurantobject/Table.java
package com.groak.groak.restaurantobject;

import com.google.firebase.Timestamp;
import com.google.firebase.firestore.DocumentReference;

import java.util.Map;

public class Table {
    private String name;
    private DocumentReference reference;
    private DocumentReference originalReference;
    private DocumentReference orderReference;
    private DocumentReference requestReference;
    private DocumentReference restaurantReference;
    private TableStatus status;
    private boolean newRequest;
    private Timestamp serveTime;

    public Table() {
        this.name = "";
        this.reference = null;
        this.originalReference = null;
        this.orderReference = null;
        this.requestReference = null;
        this.restaurantReference = null;
        this.status = TableStatus.available;
        this.newRequest = false;
        this.serveTime = null;
    }

    public Table(Map<String, Object> map) {
        this.name = (String)map.get("name");
        this.reference = (DocumentReference)map.get("reference");
        this.originalReference = (DocumentReference)map.get("originalReference");
        this.orderReference = (DocumentReference)map.get("orderReference");
        this.requestReference = (DocumentReference)map.get("requestReference");
        this.restaurantReference = (DocumentReference)map.get("restaurantReference");
        this.status = TableStatus.fromString((String)map.get("status"));
        this.newRequest = (Boolean)map.get("newRequest");
        this.serveTime = (Timestamp)map.get("serveTime");
    }

    public String getName() {
        return name;
    }
    public DocumentReference getReference() {
        return reference;
    }
    public DocumentReference getOriginalReference() {
        return originalReference;
    }
    public DocumentReference getOrderReference() {
        return orderReference;
    }
    public DocumentReference getRequestReference() {
        return requestReference;
    }
    public DocumentReference getRestaurantReference() {
        return restaurantReference;
    }
    public TableStatus getStatus() {
        return status;
    }
    public boolean isNewRequest() {
        return newRequest;
    }
    public Timestamp getServeTime() {
        return serveTime;
    }

    @Override
    public String toString() {
        return "Table{" +
                "name='" + name + '\'' +
                ", reference=" + reference +
                ", originalReference=" + originalReference +
                ", orderReference=" + orderReference +
                ", requestReference=" + requestReference +
                ", restaurantReference=" + restaurantReference +
                ", status=" + status +
                ", newRequest=" + newRequest +
                ", serveTime=" + serveTime +
                '}';
    }
}